File: android/src/main/kotlin/com/jl_dfu/jl_dfu/JlDfuPlugin.kt
package com.jl_dfu.jl_dfu

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import com.jieli.jl_bt_ota.callback.IUpgradeCallback
import com.jieli.jl_bt_ota.OTAClient
import com.jieli.jl_bt_ota.model.OTAConfigure

/**
 * JlDfuPlugin
 *
 * Implements OTA update features using the Jieli jl_bt_ota library. This plugin exposes
 * methods to start and cancel an OTA update. It assumes the application already
 * manages the Bluetooth connection and passes in a file path. Progress and status
 * events are sent back to Flutter via an EventChannel. Status strings map to
 * JL_OTAResult values defined in the iOS/Android SDKs.
 */
class JlDfuPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var progressChannel: EventChannel
    private var progressSink: EventChannel.EventSink? = null
    private var otaClient: OTAClient? = null

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "jl_dfu")
        methodChannel.setMethodCallHandler(this)
        progressChannel = EventChannel(binding.binaryMessenger, "jl_dfu_progress")
        progressChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        progressChannel.setStreamHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startOtaUpdate" -> {
                val filePath: String? = call.argument("filePath")
                if (filePath.isNullOrEmpty()) {
                    result.error("INVALID_ARGUMENT", "filePath is required", null)
                    return
                }
                // Build OTA configuration with file path
                val configure = OTAConfigure.Builder()
                    .setFilePath(filePath)
                    .build()
                otaClient = OTAClient.getDefault()
                otaClient?.startOTA(configure, object : IUpgradeCallback {
                    override fun onStartOTA() {
                        progressSink?.success(mapOf("progress" to 0.0, "status" to "preparing"))
                    }

                    override fun onProgress(progress: Float) {
                        // progress provided as 0-100; normalize to 0-1
                        val normalized = progress.toDouble() / 100.0
                        progressSink?.success(mapOf("progress" to normalized, "status" to "upgrading"))
                    }

                    override fun onStopOTA() {
                        progressSink?.success(mapOf("progress" to 1.0, "status" to "success"))
                    }

                    override fun onCancelOTA() {
                        progressSink?.success(mapOf("progress" to 0.0, "status" to "cancel"))
                    }

                    override fun onError(code: Int) {
                        progressSink?.success(mapOf("progress" to 0.0, "status" to "fail", "errorCode" to code))
                    }
                })
                result.success(null)
            }
            "cancelOtaUpdate" -> {
                otaClient?.cancelOTA()
                progressSink?.success(mapOf("progress" to 0.0, "status" to "cancel"))
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        progressSink = events
    }

    override fun onCancel(arguments: Any?) {
        progressSink = null
    }
}

File: ios/Classes/JlDfuPlugin.swift
import Flutter
import UIKit
import JL_OTALib

public class JlDfuPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "jl_dfu", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "jl_dfu_progress", binaryMessenger: registrar.messenger())
        let instance = JlDfuPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startOtaUpdate":
            if let args = call.arguments as? [String: Any],
               let filePath = args["filePath"] as? String {
                // TODO: Implement OTA update using JL_OTALib with provided file path.
                result(FlutterMethodNotImplemented)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "filePath is required", details: nil))
            }
        case "cancelOtaUpdate":
            // TODO: Implement canceling OTA update.
            result(FlutterMethodNotImplemented)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}
